src/shared-validator.cc

```diff
@@ -895,11 +895,16 @@ Result SharedValidator::OnIf(const Location& loc, Type sig_type) {
 
 Result SharedValidator::OnLoad(const Location& loc,
                                Opcode opcode,
-                               Address alignment) {
+                               Address alignment,
+                               Var memidx) {
   Result result = Result::Ok;
   MemoryType mt;
   expr_loc_ = &loc;
-  result |= CheckMemoryIndex(Var(0, loc), &mt);
+  if (options_.features.multi_memory_enabled()) {
+    result |= CheckMemoryIndex(memidx, &mt);
+  } else {
+    result |= CheckMemoryIndex(Var(0,loc), &mt);
```

Comment:
Can't we just us `CheckMemoryIndex(memidx, &mt);` in both cases?  Won't memidx above be zero anyway for the single-memory case?

Same below.

Reply:
This was kept for the same reason as OnOpcodeUint32Uint32Uint32. For load and store expressions, it is fine to only keep CheckMemoryIndex(memidx). However, for other memory instructions (memory.copy, memory.fill, etc.), only keeping CheckMemoryIndex(memidx) makes the program failed to pass some tests because the error messages are different.

Reply:
Interesting.  I'm curious to see what how the error message are different.  It could be that we simple accept the new error message and rebase the test expectations?  Can you given an example of the differences.

Reply:
Here is an example of the test output, essentially CheckMemoryIndex(memidx) attaches a location prior to the error message: 

>  test/spec/memory.txt                                                                             
>   STDOUT MISMATCH:
>   --- expected
>   +++ actual
>   @@ -22,7 +22,7 @@
>    out/test/spec/memory.wast:37: assert_invalid passed:
>      000001d: error: load/store memory 0 out of range 0
>    out/test/spec/memory.wast:41: assert_invalid passed:
>   -  error: memory variable out of range: 0 (max 0)
>   +  0000000: error: memory variable out of range: 0 (max 0)
>      0000019: error: OnMemorySizeExpr callback failed
>    out/test/spec/memory.wast:45: assert_invalid passed:
>      error: memory variable out of range: 0 (max 0)

Again, we could also rebase the test output if the change doesn't affect any existing parts of wabt.

Reply:
Yeah I think rebasing the output for this kind of change is fine.